src/mlpack/core/optimizers/parallel_sgd/parallel_sgd_impl.hpp

```diff
@@ -0,0 +1,141 @@
+/**
+ * @file parallel_sgd_impl.hpp
+ * @author Shikhar Bhardwaj
+ *
+ * Implementation of Parallel Stochastic Gradient Descent.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+#define MLPACK_CORE_OPTIMIZERS_PARALLEL_SGD_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "parallel_sgd.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+template <typename SparseFunctionType, typename DecayPolicyType>
+ParallelSGD<SparseFunctionType, DecayPolicyType>::ParallelSGD(
+    SparseFunctionType& function,
+    const size_t maxIterations,
+    const size_t batchSize,
+    const double tolerance,
+    const DecayPolicyType& decayPolicy) :
+    function(function),
+    maxIterations(maxIterations),
+    batchSize(batchSize),
+    tolerance(tolerance),
+    decayPolicy(decayPolicy)
+{ /* Nothing to do. */ }
+
+template <typename SparseFunctionType, typename DecayPolicyType>
+double ParallelSGD<SparseFunctionType, DecayPolicyType>::Optimize(
+    SparseFunctionType& function,
+    arma::mat& iterate)
+{
+  double overallObjective = 0;
+  double lastObjective = DBL_MAX;
+
+  for (size_t i = 1; i <= maxIterations; ++i){
+    overallObjective = 0;
+
+    // Get the stepsize for this iteration
+    double stepSize = decayPolicy.StepSize(i);
+
+    arma::Col<size_t> visitationOrder;
+    GenerateVisitationOrder(visitationOrder);
+
+    #pragma omp parallel
+    {
```

Comment:
I'm not sure on this one, but is it better to do `#pragma omp parallel` at a higher scope (at the beginning of the function)?  The overhead to start threads can be quite high so I am not sure if this will cause that overhead to happen every iteration.